cpp/modmesh/buffer/SimpleArray.hpp

```diff
@@ -817,6 +819,73 @@ detail::SimpleArrayMixinCalculators<A, T>::median_freq(small_vector<value_type>
     }
 }
 
+/**
+ * Perform matrix multiplication for 2D arrays.
+ * This implementation supports only 2D × 2D matrix multiplication.
+ */
+template <typename A, typename T>
+A SimpleArrayMixinCalculators<A, T>::matmul(A const & other) const
+{
+    auto athis = static_cast<A const *>(this);
+    const size_t this_ndim = athis->ndim();
+    const size_t other_ndim = other.ndim();
+
+    auto format_shape = [](A const * arr) -> std::string
+    {
+        Formatter shape_formatter;
+        if (arr->ndim() == 0)
+        {
+            shape_formatter << "()";
+        }
+        else
+        {
+            shape_formatter << "(";
+            for (size_t i = 0; i < arr->ndim(); ++i)
+            {
+                if (i > 0)
+                    shape_formatter << ",";
+                shape_formatter << arr->shape(i);
+            }
+            shape_formatter << ")";
+        }
+        return shape_formatter.str();
+    };
+
+    if (this_ndim != 2 || other_ndim != 2)
+    {
+        throw std::out_of_range(Formatter() << "SimpleArray::matmul(): unsupported dimensions: this="
+                                            << format_shape(athis) << " other=" << format_shape(&other)
+                                            << ". Only 2D x 2D matrix multiplication is supported");
+    }
+
+    const size_t m = athis->shape(0);
+    const size_t k = athis->shape(1);
+    const size_t n = other.shape(1);
+
+    if (k != other.shape(0))
```

Comment:
Compacting the logic for making the error message can improve readability, but it is not critical.  We can leave it as it for now.